TombRaiderIII/Autosplitter.cs

```diff
@@ -0,0 +1,190 @@
+        // South Pacific
+        CoastalVillage = 05,
+        CrashSite = 06,
+        MadubuGorge = 07,
+        TempleOfPuna = 08,
+        // London
+        ThamesWharf = 09,
+        Aldwych = 10,
+        LudsGate = 11,
+        City = 12,
+        // Nevada
+        NevadaDesert = 13,
+        HighSecurityCompound = 14,
+        Area51 = 15,
+        // Antarctica
+        Antarctica = 16,
+        RxTechMines = 17,
+        LostCityOfTinnos = 18,
+        MeteoriteCavern = 19,
+        // Bonus (All Secrets)
+        AllHallows = 20
+    }
+
+    /// <summary>
+    ///     Implementation of <see cref="IAutoSplitter"/> for an <see cref="AutoSplitComponent"/>'s use.
+    /// </summary>
+    internal class Autosplitter : IAutoSplitter, IDisposable
+    {
+        private const int IgtTicksPerSecond = 30;
+        private const int LevelCount = 20; // This considers Lara's Home as 0 and includes the bonus level.
+
+        private readonly HashSet<Level> _completedLevels = new HashSet<Level>(LevelCount);
+        private Level _startingLevel;
+
+        internal readonly ComponentSettings Settings = new ComponentSettings();
+        internal GameMemory GameMemory = new GameMemory();
+
+        /// <summary>A constructor that primarily exists to handle events/delegations.</summary>
+        public Autosplitter() => GameMemory.OnGameFound += Settings.SetGameVersion;
+
+        /// <summary>
+        ///     Determines the IGT.
+        /// </summary>
+        /// <param name="state"><see cref="LiveSplitState"/> passed by LiveSplit</param>
+        /// <returns>IGT as a <see cref="TimeSpan"/> if available, otherwise <see langword="null"/></returns>
+        public TimeSpan? GetGameTime(LiveSplitState state)
+        {
+            uint currentLevelTicks = GameMemory.Data.LevelTime.Current;
+            double currentLevelTime = (double)currentLevelTicks / IgtTicksPerSecond;
+
+            bool noLevelsCompleted = _completedLevels.Count != 0;
+            return noLevelsCompleted 
+                ? TimeSpan.FromSeconds(currentLevelTime)
+                : TimeSpan.FromSeconds(currentLevelTime + SumCompletedLevelTimes());
+        }
+
+        /// <summary>
+        ///     Sums completed levels' times.
+        /// </summary>
+        /// <returns>The sum of completed levels' times as a double.</returns>
+        /// <remarks>
+        ///     Because TR3's level order is variable and we must maintain compatibility with Section and NG+ runs,
+        ///     we read addresses based on <c>_completedLevels</c>.
+        /// </remarks>
+        private double SumCompletedLevelTimes()
+        {
+            var finishedLevelsTicks = 0;
+            // Add up the level times of interest from the game's memory.
+            foreach (Level completedLevel in _completedLevels)
+            {
+                int levelOffset = ((int)completedLevel - 1) * 0x2c; // - 1 because FirstLevelTimeAddress is already on the first address
```

Comment:
We should get rid of the magic const (this applies to TR2 and TR2Gold as well).

I'd drop the comment too but I don't insist on that.